### ISSUE TITLE
Not switching?

### ISSUE THREAD

**Issue by mikestecker** [jalendport/switch-php#2]
I love the idea for this but I can't seem to get it to work correctly. I originally had only PHP 7.1 installed with Homebrew and I use Valet... To test the script I went ahead and just now installed 7.2 with Homebrew then tried to switch:

```
> switch-php 72 -v -m 512M                                                                               

 👀  Verifying that Valet is installed...
Password:
 🔍  Checking which PHP versions are installed...
 🛑  Stopping Valet...
 ==>  Stopping nginx...
 ✅  Valet stopped
 🔀  Switching to php72...
 ==>  Stopping php71...
 ==>  Unlinking php71...
 ==>  Stopping php72...
 ==>  Unlinking php72...
 ==>  Linking php72...
 ==>  Starting php72...
 ✅  PHP switched
 ⚙  Starting Valet...
 ==>  Starting nginx...
 ==>  Starting dnsmasq...
 ✅  Valet started
 🎛  Configuring PHP...
 ==>  Setting PHP memory to 512M...
 ==>  Restarting PHP...
 ✅  PHP configured

You are now using PHP 7.1.14
```

Bug? Did I maybe not install 7.2 correctly? I installed 7.2 with `brew install php72 php72-igbinary php72-imagick php72-memcached`

**Comment by mikestecker** [comment]
Hmmm on further inspection, when I installed 7.2 with `brew install php72`, it gave me this:

```
The php.ini and php-fpm.ini file can be found in:
    /usr/local/etc/php/7.2/

To have launchd start php now and restart at login:
  brew services start php
Or, if you don't want/need a background service you can just run:
  php-fpm
==> Summary
🍺  /usr/local/Cellar/php/7.2.3_3: 515 files, 79MB
```

I think your script tries to start 7.2 with `brew services start php72` yet for some reason Homebrew installed it as simply `php`

**Comment by mikestecker** [comment]
Here are my versions:

```
> brew ls --versions php
php 7.2.3_3
> brew ls --versions php71
php71 7.1.14_25
```

**Comment by jalendport** [comment]
Interesting... 

I just removed php72 on my end, reinstalled it like you did, and ran my script. Worked fine on my end so will have to continue to debug what's going on here.

Can you do me a favor and send me the output of `which php` and `php -v`?

**Comment by mikestecker** [comment]
Yeah, it's very strange... here's what I just did:

```
> which php
/usr/local/opt/php71/bin/php
> php -v
PHP 7.1.14 (cli) (built: Feb  2 2018 08:42:59) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
> switch-php 72 -v -m 512M
 👀  Verifying that Valet is installed...
Password:
 🔍  Checking which PHP versions are installed...
 🛑  Stopping Valet...
 ==>  Stopping nginx...
 ✅  Valet stopped
 🔀  Switching to php72...
 ==>  Stopping php71...
 ==>  Unlinking php71...
 ==>  Stopping php72...
 ==>  Unlinking php72...
 ==>  Linking php72...
 ==>  Starting php72...
 ✅  PHP switched
 ⚙  Starting Valet...
 ==>  Starting nginx...
 ==>  Starting dnsmasq...
 ✅  Valet started
 🎛  Configuring PHP...
 ==>  Setting PHP memory to 512M...
 ==>  Restarting PHP...
 ✅  PHP configured

You are now using PHP 7.1.14

> which php               
/usr/local/opt/php71/bin/php

> php -v                  
PHP 7.1.14 (cli) (built: Feb  2 2018 08:42:59) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
```

could you add a "super" verbose to see what commands it's actually performing?

**Comment by mikestecker** [comment]
I just had a look in `/usr/local/opt/` and found this:

```
lrwxr-xr-x    1 mikestecker  admin    21 Mar 23 13:01 php -> ../Cellar/php/7.2.3_3
lrwxr-xr-x    1 mikestecker  admin    25 Mar 23 12:23 php71 -> ../Cellar/php71/7.1.14_25
lrwxr-xr-x    1 mikestecker  admin    30 Jan  2 09:28 php71-igbinary -> ../Cellar/php71-igbinary/2.0.5
lrwxr-xr-x    1 mikestecker  admin    31 Jan  2 09:50 php71-imagick -> ../Cellar/php71-imagick/3.4.3_6
lrwxr-xr-x    1 mikestecker  admin    30 Feb 14 10:23 php71-intl -> ../Cellar/php71-intl/7.1.14_25
lrwxr-xr-x    1 mikestecker  admin    32 Feb 14 10:23 php71-mcrypt -> ../Cellar/php71-mcrypt/7.1.14_20
lrwxr-xr-x    1 mikestecker  admin    33 Jan  2 09:29 php71-memcached -> ../Cellar/php71-memcached/3.0.3_3
lrwxr-xr-x    1 mikestecker  admin    35 Mar 16 09:33 php71-pdo-pgsql -> ../Cellar/php71-pdo-pgsql/7.1.14_20
drwxr-xr-x    3 root         admin    96 Jan  3 11:26 php72
lrwxr-xr-x    1 mikestecker  admin    30 Mar 23 12:12 php72-igbinary -> ../Cellar/php72-igbinary/2.0.5
lrwxr-xr-x    1 mikestecker  admin    31 Mar 23 12:12 php72-imagick -> ../Cellar/php72-imagick/3.4.3_3
lrwxr-xr-x    1 mikestecker  admin    33 Mar 23 12:12 php72-memcached -> ../Cellar/php72-memcached/3.0.3_2
```

**Comment by mikestecker** [comment]
something is definitely jacked up with my install

**Comment by jalendport** [comment]
Yes something seems weird. `which php` should always return `/usr/local/bin/php` if I'm not mistaken. Let me do a bit more research...

**Comment by jalendport** [comment]
Can you try redoing that?

I'd suggest also running:
`brew update && brew upgrade`
`brew cleanup`
`brew doctor`

Then try installing php71 and php72 again with all your extensions check `which php`...

**Comment by mikestecker** [comment]
Hmmm still the same result.

First I ran:
`brew uninstall php71 php71-igbinary php71-imagick php71-intl php71-mcrypt php71-memcached php71-pdo-pgsql`

Then `brew uninstall php72 php72-igbinary php72-imagick php72-memcached`

and did a `brew cleanup` followed by `brew doctor`

I checked the contents of `/usr/local/opt` and saw that strange `php72` directory was still there, so I manually deleted it `sudo rm -rf /usr/local/opt/php72`

Then I reinstalled 7.1 and 7.2
`brew install php71 php71-igbinary php71-imagick php71-intl php71-mcrypt php71-memcached php71-pdo-pgsql`
`brew install php72 php72-igbinary php72-imagick php72-memcached`
followed by `brew cleanup` followed by `brew doctor` again.

`brew doctor` said that `php` was not linked, so I had to do `brew link --overwrite php` since `brew link php` gave me and error `Could not symlink bin/phar`

```
> brew link --overwrite php
Linking /usr/local/Cellar/php/7.2.3_3... 698 symlinks created

> which php
/usr/local/opt/php71/bin/php

> php -v
PHP 7.1.14 (cli) (built: Feb  2 2018 08:42:59) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies

> switch-php 72 -v -m 512M
 👀  Verifying that Valet is installed...
 🔍  Checking which PHP versions are installed...
 🛑  Stopping Valet...
 ==>  Stopping nginx...
 ✅  Valet stopped
 🔀  Switching to php72...
 ==>  Stopping php71...
 ==>  Unlinking php71...
 ==>  Stopping php72...
 ==>  Unlinking php72...
 ==>  Linking php72...
 ==>  Starting php72...
 ✅  PHP switched
 ⚙  Starting Valet...
 ==>  Starting nginx...
 ==>  Starting dnsmasq...
 ✅  Valet started
 🎛  Configuring PHP...
 ==>  Setting PHP memory to 512M...
 ==>  Restarting PHP...
 ✅  PHP configured

You are now using PHP 7.1.14

> which php
/usr/local/opt/php71/bin/php
```

Here's the contents of `/usr/local/opt` after uninstalling/installing both `php71` and `php72`:
```
lrwxr-xr-x    1 mikestecker  admin    21 Mar 23 13:17 php -> ../Cellar/php/7.2.3_3
lrwxr-xr-x    1 mikestecker  admin    25 Mar 23 13:14 php71 -> ../Cellar/php71/7.1.14_25
lrwxr-xr-x    1 mikestecker  admin    30 Mar 23 13:14 php71-igbinary -> ../Cellar/php71-igbinary/2.0.5
lrwxr-xr-x    1 mikestecker  admin    31 Mar 23 13:14 php71-imagick -> ../Cellar/php71-imagick/3.4.3_6
lrwxr-xr-x    1 mikestecker  admin    30 Mar 23 13:14 php71-intl -> ../Cellar/php71-intl/7.1.14_25
lrwxr-xr-x    1 mikestecker  admin    32 Mar 23 13:14 php71-mcrypt -> ../Cellar/php71-mcrypt/7.1.14_20
lrwxr-xr-x    1 mikestecker  admin    33 Mar 23 13:14 php71-memcached -> ../Cellar/php71-memcached/3.0.3_3
lrwxr-xr-x    1 mikestecker  admin    35 Mar 23 13:14 php71-pdo-pgsql -> ../Cellar/php71-pdo-pgsql/7.1.14_20
lrwxr-xr-x    1 mikestecker  admin    30 Mar 23 13:14 php72-igbinary -> ../Cellar/php72-igbinary/2.0.5
lrwxr-xr-x    1 mikestecker  admin    31 Mar 23 13:14 php72-imagick -> ../Cellar/php72-imagick/3.4.3_3
lrwxr-xr-x    1 mikestecker  admin    33 Mar 23 13:14 php72-memcached -> ../Cellar/php72-memcached/3.0.3_2
```

**Comment by jalendport** [comment]
How about a good old `echo $PATH`?

**Comment by jalendport** [comment]
Ah you know what - I think I know what it is. It has to do with how you've set your PATH if I'm not mistaken. From the instructions PHP gives you when you install it:

```
✩✩✩✩ PHP CLI ✩✩✩✩

If you wish to swap the PHP you use on the command line, you should add the following to ~/.bashrc, ~/.zshrc, ~/.profile or your shell's equivalent configuration file:
  export PATH="$(brew --prefix homebrew/php/php71)/bin:$PATH"
```
`$(brew --prefix homebrew/php/php71)/bin` is equal to `/usr/local/opt/php71` - I bet you've set that in your `.bashrc` or whatever.

I only use `export PATH="/usr/local/bin:/usr/local/sbin:$PATH"` in my `.bashrc` and everything works fine for me. Maybe try changing that? (Don't forget to either re-source your `.bashrc` or close/reopen your terminal in order for PATH changes to take effect.)

**Comment by mikestecker** [comment]
I was thinking the same thing and after checking, I did find that line in my `.zshrc` file, but after removing it, going through the whole uninstall/reinstall process again it still seems to be having the same problem.

Here's my `$PATH` output which obviously seems a bit jacked up, there are some things in there multiple times (I removed all the `:` and added carriage returns to read it better):
```
/usr/local/sbin
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
/usr/local/mysql/bin
/opt/X11/bin=/usr/local/opt/coreutils/libexec/gnubin
/opt/X11/bin
/Users/mikestecker/node_modules/.bin
/Users/mikestecker/.composer/vendor/bin
/Users/mikestecker/Library/Python/3.6/bin=/usr/local/Cellar/coreutils/8.29/libexec/gnubin
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
/usr/local/mysql/bin
/opt/X11/bin=/usr/local/opt/coreutils/libexec/gnubin
/opt/X11/bin
/Users/mikestecker/node_modules/.bin
/Users/mikestecker/.composer/vendor/bin
/Users/mikestecker/Library/Python/3.6/bin
```

I'm going to go through my `.zshrc` and `.bashrc` files to see about cleaning this up

**Comment by mikestecker** [comment]
when I uninstall everything, `php` points to `/usr/bin/php`

doing a `ls -la` in there shows me this:
```
lrwxr-xr-x    1 root   wheel         9 Oct  8 10:37 phar -> phar.phar
-rwxr-xr-x    1 root   wheel     14823 Feb  7 18:36 phar.phar
-rwxr-xr-x    1 root   wheel  11157792 Mar 16 02:43 php
-rwxr-xr-x    1 root   wheel      6658 Feb  7 18:36 php-config
-rwxr-xr-x    1 root   wheel      4533 Jul 15  2017 phpize
```

Does that seem normal? somehow it was installed/compiled here last month...

**Comment by jalendport** [comment]
So if you close the terminal and relaunch it and then run `which php` does it still show the `/usr/local/opt/php71/bin/php` ?

If you're really interested you could always try a reboot 😬 Sorry this is taking so long to debug. Not sure if it's an issue with my script or just some environment weirdness. Another question, what version of macOS are you on?

**Comment by mikestecker** [comment]
After reloading the terminal, `which php` points to `/usr/bin/php`

I have everything uninstalled at this point. I'm running macOS 10.13.4 Beta (17E190a)

Going to reinstall everything now...

**Comment by mikestecker** [comment]
ok... after reinstalling `php71`, we are getting somewhere. Running `which php` now points to `/usr/local/bin/php`

Next is to reinstall `php72`

**Comment by mikestecker** [comment]
After installing `php72` again, I get:

```
> which php
/usr/local/bin/php
> php -v
PHP 7.1.14 (cli) (built: Feb  2 2018 08:42:59) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2018 Zend Technologies
```

Then `brew doctor` tells me:
```
brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar
Leaving kegs unlinked can lead to build-trouble and cause brews that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  php
```

Is my built-in version causing issues? A simple `brew link php` won't work:
```
Linking /usr/local/Cellar/php/7.2.3_3... 
Error: Could not symlink bin/phar
Target /usr/local/bin/phar
is a symlink belonging to php71. You can unlink it:
  brew unlink php71

To force the link and overwrite all conflicting files:
  brew link --overwrite php

To list all files that would be deleted:
  brew link --overwrite --dry-run php
```

**Comment by jalendport** [comment]
Cool cool. Now we're getting somewhere.

So yeah, you'll see the unlinked kegs thing and you don't have to worry about it. If they're unlinked that means they're not currently being used which is what we want. I have the same thing:
![image](https://user-images.githubusercontent.com/6537823/37853078-77832124-2ebb-11e8-8a69-6ece01a8ee6b.png)

So at this point, try running this:
`brew unlink php71`
`brew link php72`

And then `php -v` - fingers crossed it says PHP 7.2.3 😄

**Comment by mikestecker** [comment]
After doing `brew link --overwrite php`, it switches properly (at least manually)
```
> php -v
PHP 7.2.3 (cli) (built: Mar 21 2018 15:48:12) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.3, Copyright (c) 1999-2018, by Zend Technologies
```

```
switch-php 72 -v -m 512M
 👀  Verifying that Valet is installed...
Password:
 🔍  Checking which PHP versions are installed...
 🛑  Stopping Valet...
 ==>  Stopping nginx...
 ✅  Valet stopped
 🔀  Switching to php72...
 ==>  Stopping php71...
 ==>  Unlinking php71...
 ==>  Stopping php72...
 ==>  Unlinking php72...
 ==>  Linking php72...
 ==>  Starting php72...
 ✅  PHP switched
 ⚙  Starting Valet...
 ==>  Starting nginx...
 ==>  Starting dnsmasq...
 ✅  Valet started
 🎛  Configuring PHP...
 ==>  Setting PHP memory to 512M...
 ==>  Restarting PHP...
 ✅  PHP configured

You are now using PHP 7.2.3
```

```
switch-php 71 -v -m 512M
 👀  Verifying that Valet is installed...
 🔍  Checking which PHP versions are installed...
 🛑  Stopping Valet...
 ==>  Stopping nginx...
 ✅  Valet stopped
 🔀  Switching to php71...
 ==>  Stopping php71...
 ==>  Unlinking php71...
 ==>  Stopping php72...
 ==>  Unlinking php72...
 ==>  Linking php71...
 ==>  Starting php71...
 ✅  PHP switched
 ⚙  Starting Valet...
 ==>  Starting nginx...
 ==>  Starting dnsmasq...
 ✅  Valet started
 🎛  Configuring PHP...
 ==>  Setting PHP memory to 512M...
 ==>  Restarting PHP...
 ✅  PHP configured

You are now using PHP 7.1.14
```

So it seems like I had to do that initial force of the link then it worked...

**Comment by mikestecker** [comment]
Switching back and forth between 71 and 72 works flawlessly now. Something about the initial install, homebrew needed the `--overwrite` flag...

**Comment by jalendport** [comment]
Great. That looks like it works.

Let me see if I can try to explain what the issue seems to have been. It doesn't have to do with the `--overwrite` flag actually. Here's what was wrong: you were manually setting the version of PHP you wanted to use on your computer with `export PATH="$(brew --prefix homebrew/php/php71)/bin:$PATH"` verses using something like `export PATH="/usr/local/bin:/usr/local/sbin:$PATH"`. (I don't blame you, it's what the PHP install script says to do and it's actually quite confusing.)

What my script does (and this is the way `brew link` works, which is what my script utilizes) is basically just symlink `/usr/local/bin/php` to the actual source of the PHP version you want. So if I'm using php72, it'll look like this (I'm in the `/usr/local/bin/` folder):
`lrwxr-xr-x    1 jalendport  admin        29 Mar 23 16:59 php -> ../Cellar/php/7.2.3_3/bin/php`
If I'm using php71, it'll look like this:
`lrwxr-xr-x    1 jalendport  admin        33 Mar 23 17:09 php -> ../Cellar/php71/7.1.14_25/bin/php`

So when you first started this issue, the symlinking stuff was working fine and my script was functioning properly but you were pointing to `/usr/local/opt/php71/bin/php` which is a fixed version of PHP verses pointing to `/usr/local/bin/php` which is basically a variable pointing to whatever version of PHP you currently have linked.

So to recap, here's what would have solved your problem if we had done this at the beginning:
1. Remove `export PATH="$(brew --prefix homebrew/php/php71)/bin:$PATH"` from your `.zshrc`
2. Relaunch your terminal and `switch-php` would have worked.

But yeah it actually doesn't have to do anything with `brew link --overwrite php` because my script does that as part of the process of switching. Anyways, glad we figured it out and thanks for hanging in there during the debug process. Appreciate it... Let me know if you have any other issues!

**Comment by mikestecker** [comment]
Well, not quite actually because by the time I had discovered that `brew link --overwrite php` is what eventually solved it, I already removed `export PATH="$(brew --prefix homebrew/php/php71)/bin:$PATH"` from my .`zshrc`, reloaded the terminal and everything. I tested this... regardless it's work now haha. thanks for the script

**Comment by jalendport** [comment]
Interesting... ok well I guess I'll go ahead and close this since it seems to be resolved. Thanks again!
